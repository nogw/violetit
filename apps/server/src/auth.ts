import jwt from 'jsonwebtoken';

import { Maybe } from '@violetit/types';

import { UserModel, IUserDocument } from './modules/user/UserModel';
import { config } from './config';

export const getUser = async (token: Maybe<string>): Promise<Maybe<IUserDocument>> => {
  if (!token) return null;

  const subToken = token.substring(4);
  const decodedToken = jwt.verify(subToken, config.JWT_SECRET) as {
    id: string;
  };

  const user = await UserModel.findOne({ _id: decodedToken.id });

  if (!user) return null;

  return user;
};

export const generateJwtToken = (user: IUserDocument) => {
  return `JWT ${jwt.sign({ id: user._id }, config.JWT_SECRET)}`;
};
