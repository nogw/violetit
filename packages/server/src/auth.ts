import jwt from 'jsonwebtoken';

import { Maybe } from '@violetit/types';

import { UserModel, IUserDocument } from './modules/user/UserModel';
import { config } from './config';

export const getUser = async (token: Maybe<string>): Promise<Maybe<IUserDocument>> => {
  if (!token) return null;

  const subToken = token.substring(4);
  const decodedToken = jwt.verify(subToken, config.JWT_SECRET) as {
    id: string;
  };

  const user = await UserModel.findOne({ _id: decodedToken.id });

  if (!user) return null;

  return user;
};

export const generateJwtToken = (userId: string) => {
  return `JWT ${jwt.sign({ id: userId }, config.JWT_SECRET)}`;
};
