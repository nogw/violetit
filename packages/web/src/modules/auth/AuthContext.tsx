import React, { useState, useMemo, useCallback } from 'react';

import { Maybe } from '@violetit/types';

import { getAuthToken, updateAuthToken } from './security';

export interface AuthContextValue {
  token: Maybe<string>;
  signin: (token: Maybe<string>, cb: VoidFunction) => void;
  signout: (cb: VoidFunction) => void;
}

export const AuthContext = React.createContext<AuthContextValue>(null!);

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [userToken, setUserToken] = useState<AuthContextValue['token']>(() =>
    getAuthToken(),
  );

  const signin = useCallback<AuthContextValue['signin']>((token, cb) => {
    updateAuthToken(token);
    setUserToken(token);
    cb();
  }, []);

  const signout = useCallback<AuthContextValue['signout']>(cb => {
    setUserToken(null);
    updateAuthToken();
    cb();
  }, []);

  const value = useMemo<AuthContextValue>(
    () => ({
      token: userToken,
      signin,
      signout,
    }),
    [userToken, signin, signout],
  );

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
