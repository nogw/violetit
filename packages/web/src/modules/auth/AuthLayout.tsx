import { Navigate, Outlet, useLocation } from 'react-router-dom';
import { Fragment } from 'react';

import { Navbar } from '@/common/Navbar';
import { useAuth } from './useAuth';

export const AuthLayout = () => {
  const { pathname } = useLocation();
  const { token } = useAuth();

  if (!token) {
    return <Navigate state={{ from: pathname }} to="/auth" />;
  }

  return (
    <Fragment>
      <Navbar title="Violetit" />
      <Outlet />
    </Fragment>
  );
};
