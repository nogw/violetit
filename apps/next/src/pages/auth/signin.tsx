import { Button, Flex } from '@violetit/ui';

import { Form, useFormik, FormikProvider, FormikHelpers } from 'formik';
import { useMutation } from 'react-relay';
import { useSnackbar } from 'notistack';
import { useContext } from 'react';
import { useRouter } from 'next/router';
import * as yup from 'yup';

import { SignInMutation, SignInMutationType } from 'src/components/Auth/mutations/SignInMutation';
import { NextPageWithLayout } from 'src/relay/ReactRelayContainer';
import { InputField } from 'src/components/Formik/InputField';

import { AuthContext } from 'src/contexts/AuthContext';
import AuthLayout from 'src/layouts/AuthLayout';

type signInValues = {
  email: string;
  password: string;
};

const SignIn: NextPageWithLayout = () => {
  const [userLogin, isPending] = useMutation<SignInMutationType>(SignInMutation);

  const { closeSnackbar, enqueueSnackbar } = useSnackbar();
  const { signin } = useContext(AuthContext);
  const { push } = useRouter();

  const onSubmit = (values: signInValues, actions: FormikHelpers<signInValues>) => {
    closeSnackbar();

    userLogin({
      variables: { input: values },
      onCompleted: ({ userLogin }) => {
        if (userLogin?.error && userLogin.error.message) {
          const inputs: Array<keyof typeof values> = ['email', 'password'];

          inputs.forEach(input => {
            actions.setFieldValue(input, '', false);
            actions.setFieldTouched(input, false);
          });

          actions.setSubmitting(false);
          enqueueSnackbar(userLogin.error.message, { variant: 'error' });
          return;
        }

        enqueueSnackbar(userLogin?.success, { variant: 'success' });

        signin(userLogin?.token, () => {
          push('/');
        });
      },
    });
  };

  const formik = useFormik<signInValues>({
    initialValues: {
      email: '',
      password: '',
    },
    validateOnMount: true,
    validationSchema: yup.object().shape({
      email: yup.string().email('Invalid email').required('Email is required'),
      password: yup.string().required('Password is required'),
    }),
    onSubmit,
  });

  const { isValid, isSubmitting } = formik;

  return (
    <FormikProvider value={formik}>
      <Form>
        <Flex direction="col" className="gap-3">
          <InputField name="email" placeholder="Email" aria-required aria-label="Email" />
          <InputField name="password" placeholder="Password" type="password" aria-required aria-label="Password" />
          <Button disabled={!isValid || isPending} type="submit" aria-label="Log in">
            {isSubmitting ? 'Wait...' : 'Log in'}
          </Button>
        </Flex>
      </Form>
    </FormikProvider>
  );
};

SignIn.getLayout = page => {
  return <AuthLayout>{page}</AuthLayout>;
};

export default SignIn;
