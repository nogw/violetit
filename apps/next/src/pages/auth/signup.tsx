import { Button, Flex } from '@violetit/ui';

import { Form, useFormik, FormikProvider, FormikHelpers } from 'formik';
import { useRouter } from 'next/router';
import { useMutation } from 'react-relay';
import { useSnackbar } from 'notistack';
import { useContext } from 'react';
import * as yup from 'yup';

import { SignUpMutation, SignUpMutationType } from 'src/components/Auth/mutations/SignUpMutation';
import { NextPageWithLayout } from 'src/relay/ReactRelayContainer';
import { InputField } from 'src/components/Formik/InputField';

import { AuthContext } from 'src/contexts/AuthContext';
import AuthLayout from 'src/layouts/AuthLayout';

type signUpValues = {
  email: string;
  username: string;
  password: string;
  passwordConfirm: string;
};

const SignUp: NextPageWithLayout = () => {
  const [userRegister, isPending] = useMutation<SignUpMutationType>(SignUpMutation);

  const { closeSnackbar, enqueueSnackbar } = useSnackbar();
  const { signin } = useContext(AuthContext);
  const { push } = useRouter();

  const onSubmit = (values: signUpValues, actions: FormikHelpers<signUpValues>) => {
    closeSnackbar();

    userRegister({
      variables: {
        input: {
          email: values.email,
          username: values.username,
          password: values.password,
        },
      },
      onCompleted: ({ userRegister }) => {
        if (userRegister?.error && userRegister.error.message) {
          const inputs: Array<keyof typeof values> = ['email', 'password', 'username', 'passwordConfirm'];

          inputs.forEach(input => {
            actions.setFieldValue(input, '', false);
            actions.setFieldTouched(input, false);
          });

          actions.setSubmitting(false);
          enqueueSnackbar(userRegister.error.message, { variant: 'error' });
          return;
        }

        enqueueSnackbar(userRegister?.success, { variant: 'success' });

        signin(userRegister?.token, () => {
          push('/');
        });
      },
    });
  };

  const formik = useFormik<signUpValues>({
    initialValues: {
      username: '',
      email: '',
      password: '',
      passwordConfirm: '',
    },
    validateOnMount: true,
    validationSchema: yup.object().shape({
      username: yup.string().min(4).max(24).required('Email is required'),
      email: yup.string().email('Invalid email').required('Email is required'),
      password: yup.string().required('Password is required'),
      passwordConfirm: yup
        .string()
        .min(8)
        .oneOf([yup.ref('password')], 'Passwords must match')
        .required('Password is required'),
    }),
    onSubmit,
  });

  const { isValid, isSubmitting } = formik;

  return (
    <FormikProvider value={formik}>
      <Form>
        <Flex direction="col" className="gap-3">
          <InputField name="username" placeholder="Username" aria-required aria-label="Username" />
          <InputField name="email" placeholder="Email" aria-required aria-label="Email" />
          <InputField name="password" placeholder="Password" type="password" aria-required aria-label="Password" />
          <InputField
            name="passwordConfirm"
            placeholder="Password Confirm"
            type="password"
            aria-required
            aria-label="Password Confirm"
          />
          <Button disabled={!isValid || isPending} type="submit" variant="primary" aria-label="Create Account">
            {isSubmitting ? 'Wait...' : 'Create Account'}
          </Button>
        </Flex>
      </Form>
    </FormikProvider>
  );
};

SignUp.getLayout = page => {
  return <AuthLayout>{page}</AuthLayout>;
};

export default SignUp;
