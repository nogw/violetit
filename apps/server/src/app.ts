import Koa, { Context, Request, Response } from 'koa';
import { graphqlHTTP, OptionsData } from 'koa-graphql';
import bodyParser from 'koa-bodyparser';
import Router from '@koa/router';
import cors from '@koa/cors';

import { getContext } from './context';
import { getUser } from './auth';
import { schema } from './schema/schema';
import { GraphQLError } from 'graphql';

const app = new Koa();
const router = new Router();

const graphQLSettingsPerReq = async (req: Request, res: Response, ctx: Context): Promise<OptionsData> => {
  const user = await getUser(ctx);

  return {
    graphiql: true,
    schema,
    pretty: true,
    context: getContext({ user }),
    customFormatErrorFn: (error: GraphQLError) => {
      /* eslint-disable no-console */
      console.log(error.message);
      console.log(error.locations);
      console.log(error.stack);
      /* eslint-enable no-console */

      return {
        message: error.message,
        locations: error.locations,
        stack: error.stack,
      };
    },
  };
};

const graphQlServer = graphqlHTTP(graphQLSettingsPerReq);

router.all('/graphql', graphQlServer);

app.use(cors({ credentials: true }));
app.use(bodyParser());
app.use(router.routes());
app.use(router.allowedMethods());

export default app;
