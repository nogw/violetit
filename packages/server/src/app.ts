import Koa, { Request } from 'koa';
import { graphqlHTTP, OptionsData } from 'koa-graphql';
import bodyParser from 'koa-bodyparser';
import Router from 'koa-router';
import cors from '@koa/cors';

import { getContext } from './getContext';
import { getUser } from './auth';
import { schema } from './schema/schema';

const app = new Koa();
const router = new Router();

const graphQLSettingsPerReq = async (req: Request): Promise<OptionsData> => {
  const user = await getUser(req.header.authorization);

  return {
    graphiql: true,
    schema,
    pretty: true,
    context: getContext({ user }),
    customFormatErrorFn: ({ message, locations, stack }) => {
      console.log(message);
      console.log(locations);
      console.log(stack);

      return {
        message,
        locations,
        stack,
      };
    },
  };
};

const graphQlServer = graphqlHTTP(graphQLSettingsPerReq);

router.all('/graphql', graphQlServer);

app.use(cors({ credentials: true }));
app.use(bodyParser());
app.use(router.routes());
app.use(router.allowedMethods());

export default app;
