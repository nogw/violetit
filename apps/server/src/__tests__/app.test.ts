import request from 'supertest';

import { connectWithMongoose, clearDbAndRestartCounters, disconnectWithMongoose, sanitizeTestObject } from '../../test';
import { generateJwtToken } from '../auth';
import { createUser } from '../modules/user/fixtures/createUser';
import app from '../app';

beforeAll(connectWithMongoose);
beforeEach(clearDbAndRestartCounters);
afterAll(disconnectWithMongoose);

describe('Application', () => {
  it('should return 200 and return logged user', async () => {
    const user = await createUser();
    const token = generateJwtToken(user);

    const query = `
      query Q {
        me {
          email
          username
        }
      }
    `;

    const variables = {};

    const payload = {
      query,
      variables,
    };

    const response = await request(app.callback())
      .post('/graphql')
      .set({
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: token,
      })
      .send(JSON.stringify(payload));

    expect(response.status).toBe(200);
    expect(response.body.data.me.email).toBe(user.email);
    expect(response.body.data.me.username).toBe(user.username);
    expect(sanitizeTestObject(response.body)).toMatchSnapshot();
  });
});
