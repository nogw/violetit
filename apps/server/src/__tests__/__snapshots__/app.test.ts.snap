// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application should return 200 and return logged user 1`] = `
Object {
  "data": Object {
    "me": Object {
      "email": "user0@violetit.com",
      "username": "user#0",
    },
  },
}
`;
